### ISSUE TITLE
Cakephp Architectural Design for v4

### ISSUE THREAD

**Issue by heywhy** [cakephp/cakephp#10852]
Cakephp is a cute framework which I had always love to work with. I started getting less confident with the capacity of the framework due a project I want to work on lately and now i'm thinking of switching to laravel due to the features it has to offer and the ways you get to bind specific services (plugin like in cake).

And it offers lot of features for building a large or enterprise application at ease by offering built in queue, notification, broadcasting system, e.t.c which cake lacks.

The routing system is also a cute one I love about laravel. I will be very happy if cakephp v4 can be a great changer for the framework future.

I also thought if cake can stop using the event system to leverage some internal bindings ability to the developer, that should be a cute boost to the framework performance. And also drop the old "convention over configuration ” saying, a little by allowing the developer to have an insight about what is going on in cake internal and also being able to customise it a little.

Mind you this are just my thoughts.:smile:

**Comment by markstory** [comment]
Thanks for the feedback. We've not started on CakePHP 4 in earnest yet. There already exist plugins for queue workers ( [quesadilla](https://github.com/josegonzalez/cakephp-queuesadilla), [cakephp-queue](https://github.com/dereuromark/cakephp-queue) ) that are maintained by members of the core team.

What specific feature in Routing do you think laravel does better than CakePHP?

**Comment by inoas** [comment]
> I also thought if cake can stop using the event system to leverage some internal bindings ability to the developer, that should be a cute boost to the framework performance. And also drop the old "convention over configuration ” saying, a little by allowing the developer to have an insight about what is going on in cake internal and also being able to customise it a little.

👎 on dropping convention over configuration with the exception of singular/plural in all layers (that should become configurable with a `NullInflector` so that we can easily have all singular tables/foreign keys/entities or all plural tables/foreign keys/entities in terms of class, field and filenames). In general I love that CakePHP just sets up conventions and I can overwrite them by configuration **only if necessary, on a case by case**.

**Comment by heywhy** [comment]
@markstory the feature I love about the routing system is the ability to add middlewares to a set or particular route instead of a request going through a set of middlewares its not intended for.

I think the queue system and other set of features needed to build an enterprise or large scale apps should be built into the framework core

To be sincere markstory, I really love and appreciate your efforts on the framework.: heart:

**Comment by heywhy** [comment]
@inoas I never meant it should be dropped but instead allow the developer to have a say into the framework so as to suit his taste.

Actually I love the conventions relate to the controllers, models and view.

**Comment by cleptric** [comment]
> the ability to add middlewares to a set or particular route
This feature will be released in CakePHP 3.5 which will drop this summer. It already works but needs some final touch. 

As mark already mentioned, [quesadilla](https://github.com/josegonzalez/cakephp-queuesadilla) is a great queueing solution for CakePHP. It's definitely stable enough to use it in enterprise application, as I use it that way 🙂  
It's easy to install, has a great api and supports a lot of backends.

Committing to a certain implementation can als have its downsides. It sometimes forces the developer to build around the framework, if he want to use something else.
So keeping the core kind of "slim" and providing create plugins is IMHO a better approach.

**Comment by heywhy** [comment]
@markstory I love the ideas behind having Application.php file, I think we should be able bootstrap a lot of plugins in there instead of requiring their bootstrap and routes files, instead the plugin author provides a class with static methods which should be invoked to strap in any of the plugin's services we need..

**Comment by markstory** [comment]
> I think we should be able bootstrap a lot of plugins in there instead of requiring their bootstrap and routes files, instead the plugin author provides a class with static methods which should be invoked to strap in any of the plugin's services we need..

We used to require plugin users to manually 'bootstrap' plugins, but it resulted in a number of setup issues for plugins, which is where the options in `Plugin::load()` came from. Do you have an idea of how the class/static method approach could work?

**Comment by saeideng** [comment]
I unhappay to adding a lot of feature/plugin to core
I prefer to using standalon class/plugin

**Comment by josephzidell** [comment]
For the queueing system, what makes laravel great are their wrappers. They will provide consistent wrappers around lots of popular libraries. Install <favorite library> with composer and use the wrapper. Don't like that one and want to change. Just swap out the library, and your userland code remains the same.

**Comment by markstory** [comment]
@josephzidell That is how quesadilla works as well.

**Comment by heywhy** [comment]
# Plugin as a Service Provider

Plugins are to be assumed to be service providers instead of what they used to be.
<br>The idea behind this term is that its going to provide the developer and plugins authors
a very cute ability to change when things come in.

I think this new architectural design would be of very great performanace boost to cakephp apps.
<br>To implement this, some new conventions and how we view what plugins are, has to come in.
We create a new file in the app/config named 'services.php' containing codes similar to this

```php
<?php
return [
    'providers' => [
        \Cake\Queue\Services\QueueService::class,
        ...
    ],
    'defers' => [
        \DebugKit\Services\DebugService::class,
        ...
    ]
]
```

The idea behind the file is that, we create Service.php maybe under namespace ```Cake\Core\Service```
<br>which every service provide should extend. The class should contain this set of methods

```php
...

class Service implements ServiceInterface
{
    public function boot(\Cake\Http\MiddlewareQueue $queue, \Psr\Http\Message\ServerRequestInterface $request);
    public function bind();
    public function when(): bool;
    ...
}
```

The main aim of having this is to be able to bootstrap the registered service.
The ```when``` method is an important method in every service, it lets us decide if the framework
should keep on bootstraping the service. If ```when()``` returns true then we invoke the other 2 methods in the order ```boot($middlewareQueue, $serverRequest)``` and ```bind()```,
if the when method returns false the service doesn't go beyond that, so we move to next service in line and perform same check.

> Now that comes to the point where i said we have to stop using the Event system to leverage the internal binding capabilities

Instead of having to use ```EventManager::instance()->dispatch('Server.buildMiddleware', ...)``` in the src/Http/Server.php,
we are going to have something similar to ```Services::bootstrap($middleware, $request)```,
which loads up the app/config/services.php file and picks up the 'providers' key first, loop through all services and instantiate,
invokes the when method to decide if to keep on bootstraping the service.

## Method Summarization

### ```when()``` method
The service provider gets to check if the necessary or required necessities are met. e.g using the DebugKit\Services\DebugService

```php
...
use Cake\Core\Configure;

public function when()
{
    return Configure::read('App.debug')
    // maybe it depends on an environment variable to finalize its setup
    if (env(...)) {
        return true;
    }

    return false;
}
...
```

This is very performant in my opinion, instead of having to the developer, check that himself the service provider can leverage that itself.

### ```boot($middleware, $request)```
This should be used by services needing to inject their middlewares or changing the request(i think there is some sort of
ambiguity here, not sure if the $request object is needed here, someone can help.:smile:)
This is to inform the service that the framework is preparing to process the request, anything you want to do before the framework starts
processing the request, do it now.:smile:

```php
public function boot($middleware, $request)
{
    $middleware
        ->add(CakeLoverMiddleware::class);

    Connection::connect(Configure::read('LovingCake'));

    if (Configure::check(...)) {
        $config = array_merge($defaultConfig, Configure::read(...));
        Configure::write(..., $config);
    }
}
```

### ```bind()```
This is meant to replace the ```Plugin::load(..., ['bootstrap' => true])```, the main reason behind this new design is to reduce the number
of ```require_once``` call made by the framework. You can get to bind your global listeners in here,

```php
public function bind()
{
    EventManager::instance()->on(new ILoveCakeListener());

    // if it has routes files
    $this->enableRoutes();

    // maybe you have just a few route to add to the app
    Router::scope(['plugin' => 'CakeLover', 'path' => '/lovers'], function () {
        Router::post('/i-love-cakephp', 'CakeLovers@add')
            ->name('cakelover');

        Router::get('/get-lover/{id}', 'CakeLovers@view')
            ->whereParam(['id' => '[0-9]+'])
            ->middleware([...]);
    });
    // or maybe define a middleware group which the developer can use
    Router::middlewareGroup('web', [...]);
}
```

I know you will be surprised with the routing system up there, don't worry, i will be talking about it.:smile:
<br>Looking at the example, we've been able eliminate the need for ```require_once 'plugin_path/bootstrap.php'```


## 'defer' key
The reason i think this would be great, is the ability the consumer of the plugin has, to decide when a service comes in, by doing this;

```php
// app/config/bootstrap.php
use Cake\Core\Services;
use Cake\Http\Server;

Services::defer(\Any\Lovely\PluginService::class)
    ->untilWhen(function () {
        return Server::getRequest()->inApiContext();
    })
```

Mind you, the defered service should be added in the 'defer' key. The function gets called when the framework bootstraping the 'defer'
services needed by the app. It simple calls the callback, before invoking ```when()``` on the service,
if the callback should return false then the service doesn't need to be instanstiated, and that gives a performance boost by not accessing the
filesystem at all.

I also thought if we can eliminate the need for requiring the bootstrap file
but instead do all bindings in app/src/Application.php ```boot()``` method

# Routing Proposal

I also thought that the routing system should be very cute and expressive, just like example i gave above.
There will be a great performance boost by having 2 routes files config/routing/web.php and config/routing/api.php,
that reduces the number of routes registered when a request is coming in.

I think the community will be very glad if we can have such routing style, separating my web interface from api.

I will be very glad if we can have a routing syntax similar to;

```php
use Cake\Routing\Router;

// get verb
Router::get('/lovely-cake', 'Lovely@showHome');
// post verb
Router::post('/lovely-cake', 'Lovely@showHome');
// put verb
Router::put('/lovely-cake', 'Lovely@showHome');
// patch verb
Router::patch('/lovely-cake', 'Lovely@showHome');
// while connect matches every http verb coming through
Router::connect('/lovely-cake', 'Lovely@showHome');

Router::scope(['path' => '/cakephp'], function () {
    Router::get('/docs/{version}', 'Docs@view')
        ->whereParam(['version' => '[0-9]']);

    // 'Tutorials' matches a subnamespace in the controller namespace
    Router::post('/cake-lessons/register', 'Tutorials/Users@register')
        ->middleware([...])
        ->name('lessons:register');
});

// defining a plugin
Router::scope([
    'plugin' => 'DebugKit',
    'path' => '/debug',
    'middleware' => '...' // will be attached to all routes
], function () {
    // ...
})

// or
Router::get(...)->plugin('DebugKit');

// resource routing can be like
Router::resource('Users')
    ->only(['edit', 'add'])
    ->addRoute(['path' => '...', 'action' => 'Api/Users@...'])
    // extension gets added to all generated routes
    ->extensions('json' || ['json']);

// and so on
```

In the routing service maybe we can have this

```php
use Cake\Routing;

public function boot($middleware, $request) {
    $this->inApiContext = $request->inApiContext();
}

public function bind()
{
    // this loads routes files related to the request context
    if ($this->inApiContext) {
        return Router::loadApiRoutes();
    }
    Router::loadWebRoutes();
}
```

or maybe in the RoutingMiddleware.

I haven't figured a way to detect if the incoming request is to hit an api or web context, i need someone to help with that.:smile:


# Need For An App Command

I think we should provide an app command which can be used in doing the following by running

## ```bin/cake app optimize```
This should concatenate all routes files into their specific scope and configuration files,
like app.php, services.php e.t.c, so that we end up having

```
/config
    /cache
        app.php
        web.php
        api.php
```

All defered services should not get their routes optimized. The cache files could contain something similar to
```php
<?php
return [
    'App' => [
        // ...
    ],
    'services' => [
        'provider' => [
            // ...
        ],
        'defer' => [
            // ...
        ]
    ],
    // ...
]
```

## ```bin/cake app unoptimize```
This clears the cache files

## ```bin/cake app namespace [argument]```
The changing the project namespace with ease.

This are just my thoughts, but will be happy if they can get implemented.:smile:

**Comment by heywhy** [comment]
I think that's enough to point out what I was thinking

**Comment by slince** [comment]
I agree very much. There are many features that cake need to add or change like routing, di, ORM, templating, console and bundle system.I love doctrine and twig. 

I'm really sorry for the large number of static classes in cake.

**Comment by robertpustulka** [comment]
@slince which static classes do you have in mind?

**Comment by dakota** [comment]
@slince I'll bite :)

* Routing - There are big changes already planned, look at the road map
* DI - No thanks :) Also, Cake already has enough DI via its various service locators. If you want more dependency injection take a look at https://github.com/friendsofcake/awesome-cakephp#dependency-injection
* ORM - What's wrong with the ORM?
* Templating - What's wrong with templating?
* Console - Big changes already planned
* Bundle system - This isn't Symfony

The thing that I love about CakePHP, and that continuously draws in new users is that you don't need to know advanced, fancy pants design patterns to use it. It's incredibly easy to extend and change to suit your exact needs. It doesn't force you to adopt any specific 3rd party service or language (i.e. Twig or Blade).

**Comment by slince** [comment]
@robertpustulka 

Cache::write/read/remember; (PSR6 ?)

Plugin::load/loadAll; 

App::className;

Configure::write/read

TableRegistry::get/config

Too much static class will break the OOP principle.

**Comment by robertpustulka** [comment]
@slince 

- `Cache` class is just a *facade*. Actual cache engines are objects. You can as well type hint `CacheEngine` in your class:
```php
function __construct(CacheEngine $cache) {
    $this->cache = $cache;
    ...
    $this->cache->read('key');
}
...
new Foo(Cache::engine('default));
```

- `TableRegistry` has been refactored to use table locators.  
Use `LocatorAwareTrait`:
```php
class Foo {
    use TableLocatorTrait;

    public function bar()
    {
        $this->tableLocator()->get('Users');
    }
```
or `ModelAwareTrait`:
```php
class Foo {
    use ModelAwareTrait;

    public function bar()
    {
        $this->loadModel('Users');
    }
```


- `App::className()` is a helper method, it does not change state. It's OK to be static. Could be replaced with namespaced function in the future.

- `Configure` / `Plugin` classes - how would you see them implemented the OOP-way?

**Comment by slince** [comment]
@dakota 

- Templating     
Maybe we should provide a more friendly way for front-end developers. like "nunjucks"[http://mozilla.github.io/nunjucks/]; this is a popular  javascript template engine. And twig has the same api.

- ORM
I still need to care about details of the databases;  Is it possible that help me generate database tables based on pre-defined entities?

- DI
”service locators“ was not recomment, [https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-11-container.md#user-content-13-recommended-usage](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-11-container.md#user-content-13-recommended-usage).

Di should be supported by the cake core

**Comment by robertpustulka** [comment]
@slince you can create database table based on your schema:

```php
$schema = new TableSchema('table_name');
$schema->addColumn('foo', 'string');
$connection = ConnectionManager:get('default');

$queries = $schema->createSql($connection);
foreach ($queries as $sql) {
    $connection->execute($sql);
}

**Comment by dakota** [comment]
@slince 

- Templating
Want to use Twig, then use Twig (https://github.com/WyriHaximus/TwigView) :) CakePHP uses PHP as a templating language. PHP works very well as a templating engine, and is where it started off. Anything else is just an abstraction. I agree that sometimes you need to use something like Twig, but your PHP framework should not force you to using a specific template engine, it should simply provide a easy way to use your engine of choice. This CakePHP does incredibly well, if you want Twig you simply run `composer require wyrihaximus/twig-view` and load the plugin. If you don't want to run command line things, then use Mixer.

- ORM
Why do you want to define your database twice, or make it harder to bootstrap a new project? If you want to generate your database table, use the Migrations table.

- DI
Again, not everybody wants to use a DIC. If you do, then great, install one of the plugins here https://github.com/friendsofcake/awesome-cakephp#dependency-injection, or write your own :) DI is already supported by the Cake core

**Comment by slince** [comment]
@robertpustulka 
>  Configure / Plugin classes - how would you see them implemented the OOP-way?

Sorry i cannot understand what you mean because of my poor english.

I like ths following implementation.

### Configure

```php
class Configuration
{
      public function get($key);
      public function set($key, $value);
      public function delegate(Configuration $confiuration);
}
$confiuration1 = new Configuration('/path/or/dir/to/config');
$confiuration2 = new Configuration('/path/or/dir/to/config2');
```
- Basic usage

```php
echo $confiuration1['foo']['bar']['baz']; //Or
 $confiuration1->get('foo.bar.baz');
```

- Delegate
You can access some key in config2 like this

```php
$confiuration1->delegate($confiuration2 );
echo $confiuration1->get('foo.bar.baz.key-in-config2');
```
- Add one custome file parser

```php
Configuration::parsers()->add(new CutomFileParser());
$confiuration1->load('/path/to/config.custom_extension');
```

### Plugin

```php
interface PluginInterface
{
}

class PluginManager
{
    public function all();

    public function install(PluginInterface $plugin);

    public function uninstall();

   public static function factory();
}
```

Usage:

```php
$manager = new PluginManager();
$manager->install(new Plugin1());
$manager->install(new Plugin2());

// Or 
PluginManager::factory()->install(new Plugin3());
```

**Comment by robertpustulka** [comment]
@slince thanks, that's what I asked about.

`Configuration` example looks nice. Could you create a PR with proposed changes?

**Comment by cleptric** [comment]
The `Configuration` stuff refs #10866

**Comment by slince** [comment]
@dakota 

>  This CakePHP does incredibly well, if you want Twig you simply run composer require wyrihaximus/twig-view and load the plugin. 

Ok

> Why do you want to define your database twice, or make it harder to bootstrap a new project? If you want to generate your database table, use the Migrations table.

ORM's ultimate goal is to let you forget the database and table.  It is  flexible to generate or change the data sheet by changing the entity

> Again, not everybody wants to use a DIC. If you do, then great, install one of the plugins here https://github.com/friendsofcake/awesome-cakephp#dependency-injection, or write your own :) DI is already supported by the Cake core

Maybe I can not let you understand the advantages of di...

**Comment by slince** [comment]
@robertpustulka   The development of cake is organized, and my pr should not be passed

**Comment by josegonzalez** [comment]
Hi @slince,

While we appreciate your comments on this github issue, perhaps it is not the best venue to discuss the sorts of changes you are asking for.

If you would like to discuss this further in a way that can lead to changes in the CakePHP codebase, please join us on either the [Slack channel](http://cakesf.herokuapp.com/) or the #cakephp channel on irc.freenode.net, where we will be more than happy to discuss changes you'd like to see.

I'm logged in as `savant` on both, feel free to message me there.

Thanks!

**Comment by burzum** [comment]
> Templating

Use one of the many plugins. I think I haven't used a template engine in the past 7-10 years. The only reason to use one for me is when you don't trust the people providing the templates and don't want to process php but instead parse the template. I don't see much benefit on putting a template engine on top of what basically can be considered as one. It's just overhead. Honestly, I don't know why so many people are obsessed with using a template engine in php.

> ORM's ultimate goal is to let you forget the database and table. It is flexible to generate or change the database by changing the entity

I *really* dislike the application magically making changes to my DB... This sounds like a pretty bad idea. What if I made a mistake or there is a bug in the code and vital data gets changed / lost? I really prefer to be in charge of my DB changes via migrations instead of having the ORM doing it magically.

> Maybe I can not let you understand the advantages of di...

I assume you're talking about DI-*containers* because Cake is already using DI (pretty broad topic). I've worked on some big applications and all of them worked just fine without the use of a DI container or anything else than what CakePHP provides. I don't see a need for a DI container in the core. There are several plugins available that will add DI for those who want it to the framework.

Since this gets requested for whatever reason from time to time we could add a section about how Cake uses DI to the manual that provides links to the plugins for people who miss their container.

Actually the way DI is used in Laravel made me sick. I've spent some time looking at their Gate / Policy implementation and my impression of the code was this http://d2jmvmkdpg2gko.cloudfront.net/do-not-touch.jpg because of the way DI is used here.

**Comment by ionas** [comment]
### ORM should be close to SQL

I think the ORM should **NOT** let you forget about database details. That is exactly the bad parts about the 2.x ORM. 3.x is much closer to just a fluent sql builder that abstracts from vendor specifics for the most part and it is what I love about it. Also recently I am missing modern and ultra powerful vanilla SQL features such as Common Table Expressions or Window Functions and hiding more would not allow to have those ever in the ORM.

### Templating / TwigView

As for Templating there are different views obviously. AFAIK the author of TwigView uses Twig for all/most his templates instead of ctp. In a few projects I adopted parts of TwigView to let semi-technical users manage the view rendering. I cannot afford to go PHP there because it is inherently insecure/too powerful.

One reason to include TwigView is bake template generation. If the cakephp project adopted TwigView as a core project and used it - out of the box - foremost for cakephp/bake templates - that would be a gain. The bake templates would be vastly better to read and I guess even the participation in cakephp/bake would grow.

As a side effect people would have a reliable component to use if they required user customizable view rendering.

**Comment by dakota** [comment]
> One reason to include TwigView is bake template generation.

That's probably one of the best use cases for something like Twig. No reason why Bake can't use @WyriHaximus TwigView. He's not going anywhere soon :)

**Comment by dereuromark** [comment]
@ionas @dakota I recommended this for bake years ago :) https://github.com/cakephp/bake/issues/185#issuecomment-176844479
But also a bit off topic now here.

**Comment by markstory** [comment]
@heywhy The routing changes you've suggested add more static methods that I'm comfortable in adding. We've been trending away from statics for a number of reasons and I don't think we should reverse that decision at this point. I'm also not sold on the forced separation between 'web' and 'api' modes. I think we can achieve the same goals around separation without introducing modes into the framework.

I think the ideas you suggested for plugin classes are interesting and present an opportunity to add consistency between the burgeoning Application class and plugins. Reconciling what you've suggested with the current implementation without breaking compatibility in a significant way will be challenging.

**Comment by heywhy** [comment]
I think that's why it should fit into v4, since cake follows semver scheme so developers and plugins authors should expect breaking changes..I think making the changes public before v4 is out might help know what they should be expecting from v4 and they will be able to prepare on how to redesign the works..

**Comment by WyriHaximus** [comment]
> No reason why Bake can't use @WyriHaximus TwigView. He's not going anywhere soon :)

I'm not going anywhere soon, and even if I would stop using CakePHP, I'd still maintain it like I do with other packages I'm not using anymore.

**Comment by josegonzalez** [comment]
@heywhy For future reference on our public roadmap, as well as some reasoning behind why we chose this, please see [this](https://bakery.cakephp.org/2017/06/23/upcoming-cakephp-roadmap.html) post, which covers some of @markstory's talk at CakeFest 2017.

tl;dr We are trying to not require a ton of application changes - just removal of deprecated functionality - in order to upgrade to the latest version, as that caused quite a few issues for our users moving from 2.x to 3.x.

**Comment by heywhy** [comment]
@josegonzalez I've checked the post, I think you can give it a try in v3.6 and urge devs to work around it. Then by v4.x you can assume it has been adopted and the framework can drop fully into it. What do you think?

This can give a much more performance boost in the framework than what it used to be.

**Comment by dilab** [comment]
One cent about the queue system, It is actually more flexible to have plugins than to build it inside CakePHP core. 

I feel like the reason people complaining about the queue system is that we did not educate the plugins enough, people are not aware they are provided via the plugin system. 

We need more education on queue, broadcasting and so on. Could we probably have a section in the CakePHP book site? It will be all around using various plugins to accomplish certain tasks(queue and so on)

**Comment by burzum** [comment]
How are we going to deal with methods that return multiple different values in v4? We're using Scrutinizer and its code analyzer complains for example a lot about find()...->first(). 

Because we pass the result to another method that expects EntityInterface or a specific entity object and the analyzer then complains because first() can return an array as well according to the doc block but I think I never had that case. contain() is another one it likes to complain about because it's a combined getter/setter.

**Comment by markstory** [comment]
The get mode of contain() has been deprecated already, so it shouldn't pose an issue. The first() method can return an array when hydration is disabled.

**Comment by ADmad** [comment]
Most of the issues discussed here have now either been address, on the roadmap or have separate tickets.

* Queuing - There are well maintained plugins by core members.

* Routing - There's a separate issue for it #10643.

* DIC - It's on roadmap for Cake 4.1

* Cache - PSR 16 support has been added.

* Plugin - New `PluginInterface` has been added and plugin now have their own plugin class. `Plugin::load()/unload()` have been deprecated.

* App::className - This is appropriate use of static method to get FQCN from a short name

* Configure::write/read - There's a separate ticket #10866 kind of related to this.

* TableRegistry::get/config - Already deprecated, you have `TableLocator` now.

* Templating - There's well maintained plugin for `Twig`. Community can develop their own plugins for templating engines of their choice.

Hence closing this issue as it was more of a general discussion anyway. Please open new tickets for specific issues you feel are still not addressed adequately.
